### PR TITLE
Optional param names plots

### PR DESCRIPTION
**Pull request by DavAug**
See #1072

### CODE REVIEW

**Comment by DavAug** [comment]
Ok I thought flake8 would fail because I changed the .flake8 file. That does not seem to be the problem. The fails come, however, from places I didn't fiddle around with. Any ideas?

**Comment by MichaelClerx** [comment]
Not sure! We can have `env` and `venv` in .flake8 by the way. See what happens if you add them both in?

**Comment by DavAug** [comment]
That does not seem to work.

**Comment by MichaelClerx** [comment]
Oh wait, I just saw that the path is `./build/lib/pints/__init__.py`. I saw something similar with Myokit a few days ago.
Anyone understand what's going on here @martinjrobins @fcooper8472 ? Did pip change something?

**Comment by MichaelClerx** [comment]
Yeah seems to be a new pip issue https://github.com/pints-team/pints/issues/1119

**Comment by DavAug** [comment]
Ok! I think someone else might be better suited to tackle this problem.

File: .flake8 [hunk 1]
```diff
@@ -4,6 +4,7 @@ exclude=
   problems,
   pints/__init__.py,
   env
```
Comment (MichaelClerx) [inline .flake8:6]: Missing a comma here after `env`

File: examples/plotting/mcmc-autocorrelation.ipynb [hunk 2]
```diff
@@ -120,23 +117,29 @@
...
-      "text/plain": [
-       "<Figure size 432x432 with 3 Axes>"
-      ]
+      "text/plain": "<Figure size 432x432 with 3 Axes>",
```
Comment (MichaelClerx) [inline examples/plotting/mcmc-autocorrelation.ipynb:122]: There should be a figure here

File: pints/plot.py [hunk 3]
```diff
@@ -212,6 +221,18 @@ def histogram(samples, kde=False, ref_parameters=None, n_percentiles=None):
     n_list = len(samples)
     _, n_param = samples[0].shape
 
+    # Check parameter names
+    if parameter_names is not None:
+        if len(parameter_names) != n_param:
+            raise ValueError(
+                'Length of `parameter_names` must be same as number of'
+                ' parameters.')
+        for name in parameter_names:
```
Comment (MichaelClerx) [inline pints/plot.py:230]: This second check can go: it can be whatever `set_xlabel` accepts

File: pints/plot.py [hunk 4]
```diff
@@ -212,6 +221,18 @@ def histogram(samples, kde=False, ref_parameters=None, n_percentiles=None):
     n_list = len(samples)
     _, n_param = samples[0].shape
 
+    # Check parameter names
+    if parameter_names is not None:
```
Comment (MichaelClerx) [inline pints/plot.py:225]: I would do an `if parameter_names is None:` here that creates the default, and then `else` this code.
Then the second `if` inside the loop can go

File: pints/plot.py [hunk 5]
```diff
@@ -315,6 +347,18 @@ def trace(samples, ref_parameters=None, n_percentiles=None):
                 'All samples must have the same number of parameters.'
             )
 
+    # Check parameter names
+    if parameter_names is not None:
```
Comment (MichaelClerx) [inline pints/plot.py:351]: See above

File: pints/plot.py (hunk 4 above)
Comment (DavAug) [reply pints/plot.py:225]: I'm note sure I understand what you mean by loop here. Do you mean something like
```
if parameter_name is None:
    pass
elif len(parameter_names) != n_params:
   raise ....
```

Comment (DavAug) [reply pints/plot.py:225]: But I can revert the order in the plotting loops to 
```
if parameter_name is None:
   ... enumerate params ...
else:
   ... get names from parameter_names  ...
```

**Comment by DavAug** [comment]
I don't have a good reason to put the `parameter_names` in the middle of the arguments, other than that it made sense to me to have the argument before before reference parameter values. But maybe it's just me.
I can move it to the end if that is preferred.

File: pints/plot.py (hunk 4 above)
Comment (MichaelClerx) [reply pints/plot.py:225]: I meant getting rid of the `if` further down, when you're using the names. If you just create a `parameter_names` here you don't need to have an `if` in the plotting loop

Comment (MichaelClerx) [reply pints/plot.py:225]:
```
if parameter_names is None:
    parameter_names = <generate default names>
else:
    <check user input>

...

for i in ..
  for j in ...
     plot(..., parameter_names[0]
```

Comment (DavAug) [reply pints/plot.py:225]: I see, that makes sense!

**Comment by MichaelClerx** [comment]
@DavAug can you add "build" to the list of paths that flake8 ignores? Then I can merge this

File: pints/plot.py [hunk 6]
```diff
@@ -171,7 +171,12 @@ def function_between_points(f, point_1, point_2, padding=0.25, evaluations=20):
     return fig, axes
 
 
-def histogram(samples, kde=False, ref_parameters=None, n_percentiles=None):
+def histogram(
+        samples,
+        kde=False,
+        ref_parameters=None,
```
Comment (MichaelClerx) [inline pints/plot.py:177]: Maybe put `ref_parameters` at the end? If we're changing stuff anyway :sweat_smile: 

@chonlei @ben18785 @martinjrobins , what do you think the param order here (and in the other functions) should be?

**Comment by MichaelClerx** [comment]
> @DavAug can you add "build" to the list of paths that flake8 ignores? Then I can merge this

Once we decide on the parameter order anyway :D

File: pints/plot.py (hunk 6 above)
Comment (chonlei) [reply pints/plot.py:177]: What's wrong with `parameter_names` at the end?

Comment (MichaelClerx) [reply pints/plot.py:177]: Just trying to see if there's any particular 'best' order? Like it makes sense to put the two "parameters" arguments next to each other. And then maybe names will be used more often than ref_parameters ?

Comment (ben18785) [reply pints/plot.py:177]: Not helpful but I am completely ambivalent about this! Reckon merge it as, if we can't think of an obvious order, there likely isn't one.

Comment (MichaelClerx) [reply pints/plot.py:177]: :D OK!

In that case I'll just decide. @DavAug would you mind moving `ref_parameters` to the final position, after `parameter_names`, in the two cases we have them? After that I'm happy to merge

Comment (DavAug) [reply pints/plot.py:177]: Yep no problem.

**Comment by MichaelClerx** [comment]
Ugh, now code-cove isn't working for this PR?

**Comment by MichaelClerx** [comment]
> Run codecov
> 
>       _____          _
>      / ____|        | |
>     | |     ___   __| | ___  ___ _____   __
>     | |    / _ \ / _  |/ _ \/ __/ _ \ \ / /
>     | |___| (_) | (_| |  __/ (_| (_) \ V /
>      \_____\___/ \____|\___|\___\___/ \_/
>                                     v2.0.22
> 
> ==> Detecting CI provider
>   -> Got branch from git/hg
>   -> Got sha from git/hg
> ==> Preparing upload
> Error: Missing repository upload token
> 

@fcooper8472 do you think this was a random error, or is it to do with this being on a fork?

**Comment by MichaelClerx** [comment]
Hmmm. Seems this _should_ work https://github.com/codecov/codecov-action/issues/29

**Comment by MichaelClerx** [comment]
Pfffft. Just pulling this into a local branch instead, let's see if that works :D

**Pull request merged**